### PR TITLE
v0.1.9 updates from latest preprint: parallel hyperparameter optimisation, nested CV, matbench benchmarking code

### PR DESCRIPTION
**Pull request by ppdebreuck**
Small update to fit preset:
- dynamically creates presets in function of the number of features and sample size.
- The first layer is basically a function of the number of input features (which is the "zeroth" layer)
- If many (>2,000) samples, go from 4 to 5 layers to add complexity.

- `refit` argument to enable refitting on all data (default is True, as refitting on all data results in general to a better model)

Note: long term idea is to implement a GA, so this should somehow be temporary, or the 'fast' version.

EDIT from @ml-evs:

This PR also adds:

- nested CV and parallel hyperparameter grid searches
- improved grid presets that scale with sample and feature number
- matbench benchmarking utilities
- compatibility with precomputed data from modnet 0.1.7

### CODE REVIEW

File: modnet/models.py [hunk 1]
```diff
@@ -387,14 +371,36 @@ def fit_preset(
...
+                self.targets,
+                self.weights,
+                num_neurons=best_preset['num_neurons'],
+                n_feat=n_feat,
+                act=best_preset['act'],
+                num_classes=self.num_classes).model
+            self.n_feat = n_feat
+            self.fit(
+                data,
+                val_fraction=0.2,
```
Comment (ppdebreuck) [inline modnet/models.py:394]: should be zero, will fix this...

File: modnet/model_presets/presets.py [hunk 2]
```diff
@@ -1,33 +1,40 @@
...
 
-archs = [
-    (350, [[64], [8], [8], []]),
-    (350, [[256], [128], [8], [8]]),
-    (1000, [[256], [128], [8], []]),
-]
+    if n_feat<500: # For compounds with limited knowledge, e.g. composition only
+        n_feat_list = [60, 110]
+    else:
+        n_feat_list = [320, 520]
```
Comment (ml-evs) [inline modnet/model_presets/presets.py:12]: I think something like this might work better? We might want to be careful about someone trying to make a model from very few features too...

```suggestion
n_feat_list = (min(0.2 * n_feat, 50), max(0.5 * n_feat, 500))
```

Comment (ml-evs) [reply modnet/model_presets/presets.py:12]: (I'll push this but we can discuss)

Comment (ppdebreuck) [reply modnet/model_presets/presets.py]: ok for me !

Comment (ppdebreuck) [reply modnet/model_presets/presets.py]: or could add it to n_feat_list (if it doesn't make it to much to compute) => the chosen values (60,110,320,520) were often occurring in our tests.

File: modnet/models.py [hunk 3]
```diff
@@ -328,29 +329,11 @@ def fit_preset(
...
-        es = keras.callbacks.EarlyStopping(
-            monitor="loss",
-            min_delta=0.001,
-            patience=300,
-            verbose=verbose,
-            mode="auto",
-            baseline=None,
-            restore_best_weights=True,
-        )
-        callbacks = [rlr, es]
```
Comment (ml-evs) [inline modnet/models.py:348]: What was the reasoning for removing these?

Comment (ppdebreuck) [reply modnet/models.py:348]: After some research, learning rate scheduling is something meant for SGD optimizers. The Adam optimizer automatically reduces the learning rate dynamically, separately for each weight. So seems to be actually not necessary when using Adam. Removing early stopping was more for simplicity/better tuning (relying on epochs hyperparam instead on the callback that could be a bit arbitrary in my opinion).

File: modnet/featurizers/presets/debreuck_2020.py [hunk 4]
```diff
@@ -113,44 +114,57 @@ class DeBreuck2020Featurizer(modnet.featurizers.MODFeaturizer):
...
-        df['AtomicOrbitals|HOMO_element'] = df['AtomicOrbitals|HOMO_element'].apply(
+        df["AtomicOrbitals|HOMO_element"] = df["AtomicOrbitals|HOMO_element"].apply(
             lambda x: -1 if not isinstance(x, str) else Element(x).Z
         )
-        df['AtomicOrbitals|LUMO_element'] = df['AtomicOrbitals|LUMO_element'].apply(
+        df["AtomicOrbitals|LUMO_element"] = df["AtomicOrbitals|LUMO_element"].apply(
             lambda x: -1 if not isinstance(x, str) else Element(x).Z
         )
 
-        df = df.replace([np.inf, -np.inf, np.nan], 0)
```
Comment (ppdebreuck) [inline modnet/featurizers/presets/debreuck_2020.py:133]: We should check this throughout the whole featurization process. Do we consistently replace the nan with some value? here it is zero, but in cleaning it is -1.

Comment (ppdebreuck) [reply modnet/featurizers/presets/debreuck_2020.py:133]: (could potentially interfere with the test. at least small pre-featurized_df. But exclusively calling cleaning_df is indeed the most consitent way f doing things. Should preprocess the df in the same way for prediction.

File: modnet/featurizers/utils.py [hunk 5]
```diff
@@ -15,8 +15,8 @@ def clean_df(df):
 
     """
 
-    df = df.dropna(axis=1, how='all')
+    df = df.dropna(axis=1, how="all")
     df = df.replace([np.inf, -np.inf, np.nan], -1)
```
Comment (ppdebreuck) [inline modnet/featurizers/utils.py:19]: for a future update on MODNet: should think on what is the most appropriate way to deal with nans and inf. For now, we can leave it like that of course

File: modnet/model_presets/presets.py [hunk 6]
```diff
@@ -6,64 +6,65 @@
...
+        "Proceeding with grid search: archs: {}, batch sizes: {}, learning_rates: {}".format(
+            archs, batch_sizes, learning_rates
+        )
+    )
 
     hyperparam_presets = []
-    for bs, lr, a, e, l, act in itertools.product(
-        batch_sizes, learning_rates, archs, epochs, losses, activations
+    for a, bs, lr, e, l, act, scaler in itertools.product(
+        archs, batch_sizes, learning_rates, epochs, losses, activations, xscale
```
Comment (ppdebreuck) [inline modnet/model_presets/presets.py:67]: much easier than nested loops 👍

File: modnet/models.py [hunk 7]
```diff
@@ -62,6 +61,8 @@ def __init__(
 
         self.__modnet_version__ = __version__
 
+        if n_feat is None:
```
Comment (ppdebreuck) [inline modnet/models.py:64]: why this exactly?

File: modnet/models.py [hunk 8]
```diff
@@ -231,6 +232,12 @@ def fit(
                 )
                 loss = "categorical_crossentropy"
             else:
+                # This is a compatibility thing: we have some pre-featurized data that has differing
+                # column names that we would like to use. Most of this is single-target, so using just the
+                # one column name will always work. Otherwise this will raise a missing column error unless
+                # columns are renamed.
+                if len(self.targets_flatten) == 1:
+                    targ = list(training_data.df_targets.columns)[0]
```
Comment (ppdebreuck) [inline modnet/models.py:240]: could also rename the precomputed columns or target_names

Comment (ppdebreuck) [reply modnet/models.py:240]: maybe move this condition at the very beginning, and set targets_flatten to the columns name, will fix the next comment

File: modnet/models.py [hunk 9]
```diff
@@ -252,37 +259,43 @@ def fit(
...
-                val_data.get_target_df()[self.targets_flatten].values.transpose()
-            )
+            if len(self.targets_flatten) == 1:
+                val_y = list(
+                    val_data.get_target_df().values.transpose()
+                )
+            else:
+                val_y = list(
+                    val_data.get_target_df()[self.targets_flatten].values.transpose()
+                )
```
Comment (ppdebreuck) [inline modnet/models.py:269]: cfr before

File: modnet/models.py [hunk 10]
```diff
@@ -312,10 +323,17 @@ def fit_preset(
         presets: List[Dict[str, Any]] = None,
         val_fraction: float = 0.15,
         verbose: int = 0,
-        refit=True,
+        classification: bool = False,
+        refit: bool = True,
+        fast: bool = False,
+        nested: bool = False,
```
Comment (ppdebreuck) [inline modnet/models.py:329]: maybe add possibility to choose number of folds

File: modnet/models.py [hunk 11]
```diff
@@ -326,60 +344,122 @@ def fit_preset(
...
+            nested_val_losses = []
+            for ind, (train, test) in enumerate(splits):
+                LOG.info("Initialising split #{} preset #{}/{}: {}".format(ind + 1, i + 1, len(presets), params))
+
+                val_params = {}
+                if nested:
+                    train_data, test_data = data.split((train, test))
+                    val_params["val_data"] = test_data
+                else:
+                    val_params["val_fraction"] = 0.15
```
Comment (ppdebreuck) [inline modnet/models.py:403]:
```suggestion
                    val_params["val_fraction"] = val_fraction
```

File: modnet/models.py [hunk 12]
```diff
@@ -326,60 +344,122 @@ def fit_preset(
...
+                else:
+                    LOG.info(f"Loss from curve {val_loss:3.3f}")
+                    val_loss = np.mean(learning_curve[-20:])
+
+                nested_val_losses.append(val_loss)
+
+            val_loss = np.mean(nested_val_losses)
+
             if val_loss < min(val_losses):
                 best_model = self.model
```
Comment (ppdebreuck) [inline modnet/models.py:441]: we should in fact have a callback picking up the model with lowest validation loss

File: modnet/models.py [hunk 13]
```diff
@@ -416,8 +500,8 @@ class OR only return the most probable class.
 
 
         """
-
-        x = test_data.get_featurized_df().replace([np.inf, -np.inf, np.nan], 0)[ # prevents Nan predictions if some features are inf
+        # prevents Nan predictions if some features are inf
+        x = test_data.get_featurized_df().replace([np.inf, -np.inf, np.nan], 0)[
```
Comment (ppdebreuck) [inline modnet/models.py:504]: inconsistent with -1 in featurization

File: modnet/tests/test_model.py [hunk 14]
```diff
@@ -71,10 +71,10 @@ def test_train_small_model_multi_target(subset_moddata, tf_session):
 def test_train_small_model_presets(subset_moddata, tf_session):
```
Comment (ppdebreuck) [inline modnet/tests/test_model.py:71]: should make this test faster

**Review by ppdebreuck — COMMENTED** [review]
Thanks @ml-evs ! added some comments. Most of them are things we can add later, as for now let's have something exactly as used in the paper.

File: modnet/featurizers/presets/debreuck_2020.py (hunk 4 above)
Comment (ml-evs) [reply modnet/featurizers/presets/debreuck_2020.py:133]: Agree, should raise an issue about this. I recommend keeping it the same for now (and for 0.1.9) but we may consider leaving the NaN's for a sklearn feature scaler later on.

Comment (ml-evs) [reply modnet/featurizers/presets/debreuck_2020.py:133]: Opened #28 to track this, will resolve the comments here

File: modnet/models.py (hunk 7 above)
Comment (ml-evs) [reply modnet/models.py:64]: If you're making a model to just do `fit_preset`, you don't really want to specify `n_feat` so `None` is the most appropriate value (rather than leaving it as the default 64).

File: modnet/models.py (hunk 10 above)
Comment (ml-evs) [reply modnet/models.py]: done

**Review by ml-evs — APPROVED** [review]
